### DISCUSSION TITLE
REQ: Add TimeInForce.AT_THE_CLOSE support to IB adapter

### DISCUSSION THREAD

**Discussion by fredmonroe** [nautechsystems/nautilus_trader#1660] [Ideas]
ib doesn't neatly fit the nautilus model for MOC orders - they don't use TIF to indicate a maket on close order - its actually a different order type...

so ideally it would look like this inside the adapter (pseudo code):

if (order.type == 'market') && (order.tif == 'AT_THE_CLOSE'):
   order.type = "MOC"

details here:
https://www.interactivebrokers.com/en/trading/orders/moc.php

I'd be open to trying to do this as a PR if its hard for you guys to implement or low priority
i'm new to doing a PR but not to programming

its a little tricky because the current model assumes all 1 <--> 1 mappings for the TIF and order_type fields and this is a weird special case:

https://github.com/nautechsystems/nautilus_trader/blob/5c1c3128baf1155fd2ea38cb19a89bec2ef9bf74/nautilus_trader/adapters/interactive_brokers/execution.py#L305

**Comment by cjdsellers** [comment]
Hi @fredmonroe

I think that this is a common case for the Nautilus domain model where there are often external order types we don't represent within the framework, the set of order types we do include are fairly core/generic.

Potentially all that would be required is to translate Nautilus `MarketOrder`s with the MOC TIF to this IB order type at the system boundary.

I'm not so familiar with IB though, so maybe @rsmb7z or @benjaminsingleton would like to comment here?

**Comment by rsmb7z** [comment]
> if (order.type == 'market') && (order.tif == 'AT_THE_CLOSE'):
> order.type = "MOC"

Hi @fredmonroe,
~~This logic wouldn't achieve the intended behavior because TIF specifies the duration an order remains active and is not relevant to the order type. To convert a market order into a Market on Close (MOC) order, you need to directly set the order type to "MOC" without considering the TIF condition.~~


> Potentially all that would be required is to translate Nautilus `MarketOrder`s with the MOC TIF to this IB order type at the system boundary.

~~I will review how to support these translations without confusing them with TIF, as the conversion of order types automatically happens at the broker. For instance, a Limit order would convert to a Market order right before close in this case, and this conversion has no direct context with TIF.~~

Edit: My bad, confused with something else.

**Reply by fredmonroe** [reply]
I agree that MOC seems more like a distinct order type - and it would make the code a lot easier in the ib adapter if nautilus listed it as an order type. 

this is where i got the idea that nautilus implemented MOC via combo of market_order order type + TIF
[nautilus docs on order types](https://docs.nautilustrader.io/api_reference/model/orders.html)

 _"A Market-On-Close (MOC) order can be represented using a time in force of AT_THE_CLOSE"_ 
 
 In any case I'm grateful for you taking a look, if I can help in some way please let me know

**Reply by fredmonroe** [reply]
Just another little note, I implemented the conversion originally proposed and was able to get MOC orders into IB
But - I agree with @rsmb7z - i'm not sure my original proposal makes sense and i think MOC should be a distinct order type that nautilus supports. The code would be a lot clearner for the conversion in the adapter and his arguments make sense to me.

just fyi for IB - MOC orders need a TIF of DAY - not the default nautilus uses of GTC

my implementation is pretty gross, i have to create a  new marketorder from scratch because i can't modify fields on the Order class (they are read only) - so i don't think you guys will want it, but you are welcome to it if you do - and it gets me up and running for now

https://github.com/nautechsystems/nautilus_trader/compare/develop...splatai:nautilus_trader:MOC

**Comment by rsmb7z** [comment]
Hi @fredmonroe 
I had a misunderstanding. It is indeed relevant to TIF/ActiveStartTime, but with the exception that it is a special order type, unlike Market on Open. I have pushed the support for MOC and LOC in #1663.

**Reply by fredmonroe** [reply]
Thank you! will test it out right away

**Reply by fredmonroe** [reply]
works great for me so far in IBKR paper
